Ruby Capstone- Catalog Of Things

ClintonYounge opened the pull request:
# Updates and Added features

 **User interface:**
 - Created a main.rb file that will serve as your console app entry-point.
 - Present the user with a list of options to perform.
 - Let users choose an option.
 - Have a way to quit the app.

**The following main options are available in the app:**
- Books
- Music Album
- Games
- Exit

**The sub options are**
- List all books
- List all music albums
- List all movies
- List of games
- List all genres 
- List all labels
- List all authors
- Add a book
- Add a music album
- Add a game
- Go back


 **Database:**
- We have a schema.sql file with tables that will be analogical to the structure of the classes in our app.

**Tests  and Storage:**
- Unit tests for all implemented methods.
- All data are preserved by saving collections in .json files.

ClintonYounge committed:
Add prerequisite code

- Add menu structure

mpa-mxiang committed:
autofix linter errors

mpa-mxiang committed:
fix item initialization

mpa-mxiang committed:
improve item archive

ClintonYounge committed:
Fix linters

ClintonYounge committed:
Merge pull request #45 from ClintonYounge/Prerequisites/Group-Tasks

Add prerequisite code

ClintonYounge committed:
Update main item class

ClintonYounge committed:
Run linters

ClintonYounge committed:
Merge pull request #46 from ClintonYounge/Prerequisites/Group-Tasks

Update main item class

mpa-mxiang committed:
Merge pull request #47 from ClintonYounge/dev

Merge to synchronize with dev

mpa-mxiang committed:
delete extra archive method

mpa-mxiang committed:
Create genre.rb

mpa-mxiang committed:
Create music_album.rb

sylvester-wamaya committed:
Initialize label class

sylvester-wamaya committed:
Add add_item method

sylvester-wamaya committed:
Fix linter errors

sylvester-wamaya committed:
Merge pull request #48 from ClintonYounge/add-label

Add label class

sylvester-wamaya committed:
Initialize book class

sylvester-wamaya committed:
Initialize book class

sylvester-wamaya committed:
Overirde can_be_archived class.

ClintonYounge committed:
Add games methods

ClintonYounge committed:
Merge branch 'dev' into teammate-3/add-classes

mpa-mxiang committed:
Merge pull request #49 from ClintonYounge/tm2-music

Teammate 2 - MusicAlbum&Genre

mpa-mxiang committed:
Merge branch 'dev' into teammate-3/add-classes

ClintonYounge committed:
Merge pull request #50 from ClintonYounge/teammate-3/add-classes

Add games methods

sylvester-wamaya committed:
Fix errors in initializing Book

mpa-mxiang committed:
Create musicManager.rb

mpa-mxiang committed:
autofix linter errors

mpa-mxiang committed:
change file name to meet snake_case

mpa-mxiang committed:
Create genre_spec.rb

mpa-mxiang committed:
Create music_album_spec.rb

mpa-mxiang committed:
autofix linter errors

mpa-mxiang committed:
try to improve item and music classes

mpa-mxiang committed:
delete unrelated part from this pr

mpa-mxiang committed:
add date in item

mpa-mxiang committed:
Merge branch 'dev' into music-unit-tests

mpa-mxiang committed:
Merge pull request #51 from ClintonYounge/tm2-music

Music Manager class - available options

mpa-mxiang committed:
add save and load data for music albums and genres

mpa-mxiang committed:
call save and load methods

mpa-mxiang committed:
call save genres methods and autofix linter errors

sylvester-wamaya committed:
Initialize BookHandler class

mpa-mxiang committed:
Create schema.sql

mpa-mxiang committed:
call music manager methods in app

sylvester-wamaya committed:
Add method to add book, list books and labels

ClintonYounge committed:
Implement author and games storage

mpa-mxiang committed:
Merge pull request #52 from ClintonYounge/music-unit-tests

Music unit tests

mpa-mxiang committed:
Merge pull request #53 from ClintonYounge/music-data-preservation

Music data preservation

mpa-mxiang committed:
Merge pull request #54 from ClintonYounge/music-tables

Music database

sylvester-wamaya committed:
Implement iterract with books on terminal

sylvester-wamaya committed:
Implement iterract with books on terminal

sylvester-wamaya committed:
Fix linters

mpa-mxiang committed:
Merge branch 'dev' into book

mpa-mxiang committed:
Merge branch 'book' of https://github.com/ClintonYounge/Catalog-Of-Things-Ruby-Capstone into book

mpa-mxiang committed:
Merge branch 'dev' into book

sylvester-wamaya committed:
Fix linters

mpa-mxiang committed:
comment save genres method

mpa-mxiang committed:
Merge branch 'book' of https://github.com/ClintonYounge/Catalog-Of-Things-Ruby-Capstone into book

mpa-mxiang committed:
autofix linter error with comment

mpa-mxiang committed:
Merge pull request #55 from ClintonYounge/tm2-music

Link options

mpa-mxiang committed:
Merge branch 'dev' into teammate-3/data-preservation

sylvester-wamaya committed:
Fix linters

mpa-mxiang committed:
autofix space issue

mpa-mxiang committed:
Merge branch 'teammate-3/data-preservation' of https://github.com/ClintonYounge/Catalog-Of-Things-Ruby-Capstone into teammate-3/data-preservation

mpa-mxiang committed:
comment unrelated save genre method issue

sylvester-wamaya committed:
Fix code conflicts

mpa-mxiang committed:
Merge branch 'dev' into book

mpa-mxiang committed:
comment unrelated save genres method issue

ClintonYounge committed:
Update app

mpa-mxiang committed:
move save genre method to avoid test fail and autofix extra blank line

mpa-mxiang committed:
Merge pull request #58 from ClintonYounge/tm2/test-vs-data

Music Genre save method and unit test conlict

mpa-mxiang committed:
Merge branch 'dev' into teammate-3/data-preservation

mpa-mxiang committed:
Merge branch 'dev' into book

ClintonYounge committed:
Update item class

ClintonYounge committed:
Update files

ClintonYounge committed:
Update item class

ClintonYounge committed:
Run linters

ClintonYounge committed:
Merge pull request #57 from ClintonYounge/teammate-3/data-preservation

Implement author and games storage

ClintonYounge committed:
Merge pull request #56 from ClintonYounge/book

Add Book class and its methods

ClintonYounge committed:
Debug program

Co-Authored-By: Jiachun Xiang <jcxiang1230@gmail.com>
Co-Authored-By: Sylvester Wamaya <sylvesterwamaya@gmail.com>

mpa-mxiang committed:
autofix linter errors

Co-Authored-By: Clinton Younge <clintonyounge18@gmail.com>

ClintonYounge committed:
Update debug code

ClintonYounge committed:
Merge branch 'debug-branch' of https://github.com/ClintonYounge/Catalog-Of-Things-Ruby-Capstone into debug-branch

sylvester-wamaya committed:
Add save book mehtod

mpa-mxiang committed:
fix class has too much lines issue

mpa-mxiang committed:
Merge pull request #59 from ClintonYounge/debug-branch

Debug program

mpa-mxiang committed:
delete no reference option method calls

sylvester-wamaya committed:
Add load books method

sylvester-wamaya committed:
Add save label to json

sylvester-wamaya committed:
Implement storage, retreival of data in json file

ClintonYounge committed:
Remove json files

ClintonYounge committed:
Merge pull request #60 from ClintonYounge/debug-branch

Debug branch

mpa-mxiang committed:
Create genre_spec.rb

mpa-mxiang committed:
Create music_album_spec.rb

mpa-mxiang committed:
make the name same without confusion

sylvester-wamaya committed:
Fix linters

mpa-mxiang committed:
add a new music album for test

mpa-mxiang committed:
autofix linters

ClintonYounge committed:
Add game and author tests

ClintonYounge committed:
Add game methods test

ClintonYounge committed:
Add game storage tests

mpa-mxiang committed:
Merge pull request #62 from ClintonYounge/tm2/get-tests-back

Get tests back

ClintonYounge committed:
Run linters

sylvester-wamaya committed:
Merge pull request #61 from ClintonYounge/book-storage

Implement book storage

ClintonYounge committed:
Merge pull request #63 from ClintonYounge/teammate-3/unit-tests

Teammate 3/unit tests

mpa-mxiang committed:
Revert "Teammate 3/unit tests"

mpa-mxiang committed:
make item and music album back to work version

mpa-mxiang committed:
autofix linters

mpa-mxiang committed:
add four test files back

mpa-mxiang committed:
autofix linter errors

mpa-mxiang committed:
Merge pull request #64 from ClintonYounge/revert-63-teammate-3/unit-tests

Revert "Teammate 3/unit tests"

ClintonYounge committed:
Update to functioning code

ClintonYounge committed:
Fix linter errors

ClintonYounge committed:
Restructure books code

ClintonYounge committed:
Update book and label test structure

mpa-mxiang committed:
add genre item

ClintonYounge committed:
Update sql

ClintonYounge committed:
Improve UI by adding spaces

mpa-mxiang committed:
fix item count issue

ClintonYounge committed:
Fix typo error

ClintonYounge committed:
Update UI to add white space

mpa-mxiang committed:
delete unnecessary puts

mpa-mxiang committed:
autofix linters

mpa-mxiang committed:
fix extra end issue

mpa-mxiang committed:
delete already deleted cases

ClintonYounge committed:
Merge pull request #66 from ClintonYounge/branch/debug

Branch/debug

mpa-mxiang committed:
Merge branch 'dev' into genre-add

mpa-mxiang committed:
fix linter errors with too much lines in create method

mpa-mxiang committed:
fix genre storage issue

mpa-mxiang committed:
sychronize as dev version

mpa-mxiang committed:
autofix linter error

mpa-mxiang committed:
add save genre method in app

mpa-mxiang committed:
rename save genre method name and add it in main

mpa-mxiang committed:
delete save genres method from app

ClintonYounge committed:
Restructure methods save calls

ClintonYounge committed:
Update readme

mpa-mxiang committed:
Merge pull request #67 from ClintonYounge/genre-add

Genre add

ClintonYounge committed:
Update README file

tresorsawasawa commented on item.rb:
```diff
@@ -0,0 +1,28 @@
+require 'securerandom'
+require 'date'
+
+class Item
+  attr_accessor :label, :genre, :author
+  attr_reader :id, :published_date, :archived
+
+  def initialize(published_date = nil, id: nil)
+    @id = id || SecureRandom.uuid
+    @label = nil
+    @genre = nil
+    @author = nil
+    @published_date = published_date
+    @archived = false
+    move_to_archive
+  end
+
+  def can_be_archived?
+    return false if @published_date.nil?
+
+    published_date = DateTime.parse(@published_date)
+    published_date < DateTime.now - 30
+  end
+
+  def move_to_archive
+    @archived = can_be_archived?
+  end
```
- [ ] As required for the project to correctly math all associations (1-to-many relationships) between classes, kindly ensure that the `Item` class has one `Genre`, `Author`, and `Label`. If you need a refresher about associations read [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships). For example:
```ruby
class Item
    attr_reader :label, :genre, :author
   .
   .
   .
   # Same example to each one(i.e. label and author)
   def genre=(genre)
      @genre = genre
      genre.items.push(self) unless genre.items.include?(self)
    end
    ...
end
```

tresorsawasawa replied:
- [ ] As required for the project the Item class to have a `move_to_archive()` method which  reuses `can_be_archived?`method, changing the archived property to true if the result of the `can_be_archived?()` method is true, and do nothing if the result of the `can_be_archived?()` method is false, kindly ensure that you also update this to match the expected result.
```ruby
# example
def move_to_archive
      if can_be_archived?
        @archived = true
      end
    end
```

tresorsawasawa commented on item.rb:
```diff
@@ -0,0 +1,28 @@
+    move_to_archive
+  end
+
+  def can_be_archived?
+    return false if @published_date.nil?
+
+    published_date = DateTime.parse(@published_date)
+    published_date < DateTime.now - 30
+  end
```
- [ ] As required for the project the Item class to have a `can_be_archived?()` method  returning true if `published_date` is older than 10 years, otherwise, to return false, I think it should better if you update the above to match the expected result.
```ruby
# example
def can_be_archived?
    current_year = Time.now.year
    publication_year = @published_date.year

    if current_year - publication_year > 10
      return true
    else
      return false
    end
  end
```

tresorsawasawa commented on spec/author_spec.rb:
```diff
@@ -0,0 +1,15 @@
+require_relative '../author'
+
+describe Author do
+  before :each do
+    @author = Author.new('Clinton', 'Younge')
+  end
+
+  it 'should have a first name' do
+    expect(@author.first_name).to eq('Clinton')
+  end
+
+  it 'should have a last name' do
+    expect(@author.last_name).to eq('Younge')
+  end
```
- [ ] As required for the project all public methods in all classes to have unit tests, kindly ensure that you add tests to the `#add_item()` method in the `Author` class

tresorsawasawa commented on spec/music_album_spec.rb:
```diff
@@ -0,0 +1,20 @@
+      music_album = MusicAlbum.new(on_spotify: true)
+      expect(music_album.on_spotify).to eq(true)
+    end
+  end
+end
```
- [ ] As required for the project all public methods in all classes to have unit tests, kindly ensure that you also add tests to the following class(es) and their public methods.
  * `Item` class and its `#can_be_archived?()` method

tresorsawasawa reviewed (changes requested):
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @ClintonYounge, @mpa-mxiang, @sylvester-wamaya 
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

![almost-there](https://user-images.githubusercontent.com/57562869/143511358-9cde7daf-4ad0-4db2-8f15-18b54288ff91.gif)

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- All tests are passing :heavy_check_mark:
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
- Follow Gitflow :heavy_check_mark: 

## Required changes :recycle: :recycle: 

- [ ] As required for the project, kindly ensure that you record a video with the following features:
  * It should present the features of the project.
  * Its maximum length is 5 minutes.
  * It should contain a clear and objective description of the project in professional English.
  * All the members in the video should appear in the video and talk about the project.
  * It should contain a highlight about a piece of the code you find interesting or you are proud of making
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ClintonYounge committed:
Restructure Item class

sylvester-wamaya committed:
Add label tests

sylvester-wamaya committed:
Add book tests

sylvester-wamaya committed:
Date.parse published date

sylvester-wamaya committed:
Fix linter errors.

mpa-mxiang committed:
move genres items count back

mpa-mxiang committed:
autofix linter errors

ClintonYounge committed:
Restructure Game and Author methods

Restructured the list game, list authors and the overall game and author classes to properly establish the relationships.

ClintonYounge committed:
Restructure save methods

ClintonYounge committed:
Merge branch 'teammate-3/restructure' of https://github.com/ClintonYounge/Catalog-Of-Things-Ruby-Capstone into teammate-3/restructure

ClintonYounge committed:
Update game and author tests

ClintonYounge committed:
Add item tests

ClintonYounge committed:
Update game storage tests

ClintonYounge committed:
Update book can_be_archived? method

mpa-mxiang committed:
Merge pull request #69 from ClintonYounge/move-genres-items-back

move genres items count back

ClintonYounge committed:
Fix linters

The github linter is showing an error that doesn't exist. Will try to fix by changing around some code.

ClintonYounge committed:
Implement suggested fix

Introduce error to replicate bug.

ClintonYounge committed:
Run linters

ClintonYounge committed:
Update item.rb to fix linter

ClintonYounge committed:
Delete item.rb

ClintonYounge committed:
Add item.rb file

ClintonYounge committed:
Refactor genre and albums code

Refactored genre and album to establish necessary relationship.

ClintonYounge committed:
Update music and genre tests

ClintonYounge committed:
Update UI

ClintonYounge committed:
Update Readme

ClintonYounge committed:
Fix linter errors

ClintonYounge committed:
Merge pull request #70 from ClintonYounge/teammate-3/restructure

Teammate 3/restructure

afanimartin reviewed (approved):
Hi team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- More methods added to the Item and Author classes as requested
- Tests added for the new methods added

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

## Find me on slack
### Martin Afani

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
